src/main/java/com/ldtteam/blockout/Loader.java

```diff
@@ -29,92 +30,90 @@
 {
     private static final DocumentBuilderFactory DOCUMENT_BUILDER_FACTORY = DocumentBuilderFactory.newInstance();
 
-    private static final Map<String, Constructor<? extends Pane>> paneConstructorMap = new HashMap<>();
+    private static final Map<ResourceLocation, Function<PaneParams,? extends Pane>> paneFactories = new HashMap<>();
     static
     {
-        register("view", View.class);
-        register("group", Group.class);
-        register("scrollgroup", ScrollingGroup.class);
-        register("list", ScrollingList.class);
-        register("text", Text.class);
-        register("button", ButtonVanilla.class);
-        register("buttonimage", ButtonImage.class);
-        register("label", Text.class); // TODO: remove, but we don't want to deal with xml changes now
-        register("input", TextFieldVanilla.class);
-        register("image", Image.class);
-        register("imagerepeat", ImageRepeatable.class);
-        register("box", Box.class);
-        register("itemicon", ItemIcon.class);
-        register("switch", SwitchView.class);
-        register("dropdown", DropDownList.class);
-        register("overlay", OverlayView.class);
-        register("gradient", Gradient.class);
-        register("zoomdragview", ZoomDragView.class);
-        register("treeview", TreeView.class);
+        register("view", View::new);
+        register("group", Group::new);
+        register("scrollgroup", ScrollingGroup::new);
+        register("list", ScrollingList::new);
+        register("text", Text::new);
+        register("button", Button::new);
+        register("buttonimage", Button::new);
+        register("toggle", ToggleButton::new);
+        register("label", Text::new);
+        register("input", TextFieldVanilla::new);
+        register("image", Image::new);
+        register("imagerepeat", ImageRepeatable::new);
+        register("box", Box::new);
+        register("itemicon", ItemIcon::new);
+        register("switch", SwitchView::new);
+        register("dropdown", DropDownList::new);
+        register("overlay", OverlayView::new);
+        register("gradient", Gradient::new);
+        register("zoomdragview", ZoomDragView::new);
+        register("treeview", TreeView::new);
     }
 
-    private Loader()
+    /** The maximum numer of cached documents */
+    private static final int CACHE_CAP = 20;
+
+    /** A map to store the parsed documents. Retains data based on a priority */
+    private static final Map<ResourceLocation, Tuple<Integer,PaneParams>> parsedCache = Collections.synchronizedMap(new HashMap<ResourceLocation, Tuple<Integer, PaneParams>>()
     {
```

Comment:
do we need a synchronized map here? Not better to use a caffeeine cache then?

Reply:
Not completely sure, but because windows are ``open()``ed right after they're constructed I think it does have to be synced.